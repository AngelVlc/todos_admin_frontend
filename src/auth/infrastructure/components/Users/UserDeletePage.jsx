import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useParams, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetUserByIdUseCase,
  DeleteUserByIdUseCase,
} from "../../../application/users";

export const UserDeletePage = () => {
  const [user, setUser] = useState(null);
  const { useCaseFactory } = useContext(AppContext);
  let { userId } = useParams();
  let history = useHistory();

  const getUser = useCallback(async () => {
    const getUserByIdUseCase = useCaseFactory.get(GetUserByIdUseCase);
    const user = await getUserByIdUseCase.execute(userId);

    setUser(user);
  }, [userId, useCaseFactory]);

  useEffect(() => {
    getUser();
  }, [userId, getUser]);

  const deleteUser = async () => {
    const deleteUserByIdUseCase = useCaseFactory.get(DeleteUserByIdUseCase);

    if (await deleteUserByIdUseCase.execute(userId)) {
      history.push("/users");
    }
  };

  return (
    <>
      {user && (
        <div className="container">
          <h3 className="title">
            {user.isAdmin
              ? `Delete admin user '${user.name}'?`
              : `Delete user ${user.name}?`}
          </h3>
          <nav className="breadcrumb" aria-label="breadcrumbs">
            <ul>
              <li>
                <Link to={`/`}>Home</Link>
              </li>
              <li>
                <Link to={`/users`}>Users</Link>
              </li>
              <li>
                <Link to={`/users`}>Users</Link>
              </li>
              <li>
                <Link aria-current="page" to={`/users/${userId}/edit`}>
                  {user.name}
                </Link>
              </li>
              <li className="is-active">
                <Link
                  aria-current="page"
                  to={`/users/${userId}/delete`}
                >{`Delete ${user.name}`}</Link>
              </li>
            </ul>
          </nav>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="yes"
              onClick={() => deleteUser()}
            >
              YES
            </button>
            <button
              className="button"
              data-testid="no"
              onClick={() => history.goBack()}
            >
              NO
            </button>
          </div>
        </div>
      )}
    </>
  );
};
