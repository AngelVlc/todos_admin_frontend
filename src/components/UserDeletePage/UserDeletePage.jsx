import React, { useEffect, useContext, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { AppContext } from '../../contexts/AppContext';
import { doGet, doDelete } from '../../helpers/api';

export const UserDeletePage = () => {
  const { auth, requestsDispatch } = useContext(AppContext)
  const [user, setUser] = useState(null);
  let { userId } = useParams();
  let history = useHistory();

  useEffect(() => {
    const getUser = async () => {
      const res = await doGet(`users/${userId}`, auth.info.token, requestsDispatch)
      let title = `Delete user ${res.name}`
      if (res.isAdmin) {
        title = `Delete admin user '${res.name}'?`
      }
      const info = {
        ...res,
        title: title
      }
      setUser(info);
    }
    if (auth.info) {
      getUser();
    }
  }, [userId, auth.info, requestsDispatch]);

  const deleteUser = async () => {
    await doDelete(`users/${userId}`, auth.info.token, requestsDispatch)
    history.goBack();
  }

  return (
    <>
      {user &&
        <div className="container">
          <h3 className="title">{user.title}</h3>
          <div className="buttons">
            <button className="button is-danger" data-testid="yes" onClick={() => deleteUser()}>YES</button>
            <button className="button" data-testid="no" onClick={() => history.goBack()}>NO</button>
          </div>
        </div>
      }
    </>
  );
}
