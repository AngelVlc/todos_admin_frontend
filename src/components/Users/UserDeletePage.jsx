import React, { useEffect, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import axios from 'axios';

export const UserDeletePage = () => {
  const [user, setUser] = useState(null);
  let { userId } = useParams();
  let history = useHistory();

  useEffect(() => {
    const getUser = async () => {
      const res = await axios.get(`users/${userId}`);
      let title = `Delete user ${res.data.name}`;
      if (res.data.isAdmin) {
        title = `Delete admin user '${res.data.name}'?`;
      }
      const info = {
        ...res.data,
        title: title
      };
      setUser(info);
    }
    getUser();
  }, [userId]);

  const deleteUser = async () => {
    await axios.delete(`users/${userId}`);
    history.push('/users');
  }

  return (
    <>
      {user &&
        <div className="container">
          <h3 className="title">{user.title}</h3>
          <div className="buttons">
            <button className="button is-danger" data-testid="yes" onClick={() => deleteUser()}>YES</button>
            <button className="button" data-testid="no" onClick={() => history.goBack()}>NO</button>
          </div>
        </div>
      }
    </>
  );
}
