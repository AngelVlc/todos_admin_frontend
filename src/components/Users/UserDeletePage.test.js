import { render, cleanup, fireEvent, waitFor } from '@testing-library/react'
import { UserDeletePage } from './UserDeletePage'
import { AppContext } from '../../contexts/AppContext'
import axios from 'axios';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('axios');
const mockHistoryGoBack = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryGoBack,
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouter = (component, isAdmin) => {
    axios.get.mockResolvedValue(
        {
            data:
            {
                id: 2,
                name: 'user',
                isAdmin: isAdmin
            }
        }
    );
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/users/2/delete`]}>
                    <Route path="/users/:userId/delete">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot when the user is not an admin', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouter(<UserDeletePage />, false);
        fragment = asFragment;
    });
    expect(fragment(<UserDeletePage />)).toMatchSnapshot();
});

it('should match the snapshot when the user is an admin', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouter(<UserDeletePage />, true);
        fragment = asFragment;
    });
    expect(fragment(<UserDeletePage />)).toMatchSnapshot();
});

it('should cancel the deletion', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<UserDeletePage />, false);
    });

    await waitFor(() => {
        fireEvent.click(container.getByTestId('no'));
    })

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should delete a user', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<UserDeletePage />, false);
    });

    axios.delete.mockResolvedValue({data:{}});

    await waitFor(() => {
        fireEvent.click(container.getByTestId('yes'));
    })

    expect(axios.delete.mock.calls.length).toBe(1);
    expect(axios.delete.mock.calls[0][0]).toBe('users/2');
    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/users');
    mockHistoryPush.mockClear();
});
