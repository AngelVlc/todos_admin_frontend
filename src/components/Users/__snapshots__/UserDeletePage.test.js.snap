// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot when the user is an admin 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Delete admin user 'user'?
    </h3>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match the snapshot when the user is not an admin 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Delete user user
    </h3>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;
